proposals/stack-switching/Explainer.md

```diff
@@ -94,6 +94,10 @@ We add two new continuation heap types and their subtyping hierachy:
     - and `C.types[$ct2] = cont $ft2`
     - and `C.types[$ft2] = [t2*] -> [te2*]`
     - and `te2* <: t*`
+   
+### Execution
+
+The same tag may be used simultaneously by `throw`, `suspend`, `switch`, and their associated handlers. When searching for a handler for an event, only handlers for the matching kind of event are considered, e.g. only `(on $e $l)` handlers can handle `suspend` events and only `(on $e switch)` handlers can handle `switch` events. The handler search continues past handlers for the wrong kind of event, even if they use the correct tag.
```

Comment:
Hm, I understood the motivation, but this would be unfortunate in that it destroys the desugaring equivalence for switch.

Reply:
The desugaring already requires generating new tags for every combination of symmetric switching tag (to make sure the new `suspend`s go to the right place) and participating continuation type (to make sure the new `(on $e $l)` handlers can receive all the various continuation types to be resumed), so I don't think this changes anything in that regard.

And unless we want to start using tag attributes to differentiate tags for different uses, we have no way to prevent tags from being used for different uses simultaneously and we still have to determine the semantics in such cases. Is there some other semantics you would prefer?

Reply:
The tag multiplication is merely due to a typing limitation. It can be avoided by assuming a relatively minor typing extension, namely allowing tag types with local type variables, as in:
```
tag $t : α* (ref (cont α* (ref $ct) → t'*)) → t*
```
This is a standard form of a simple existential type, a well-known device to e.g. type closures.

But we don't even need that, since we have it already! It's the whole reason why we needed cont.bind. That allows unifying continuation types by closure formation, and is crucial in various scenarios we have looked at. The same idea works here: we can desugar `switch` into a combination of `cont.bind` + `suspend`, such that the actual desugared switch tag has no parameters except for the target continuation, and likewise that continuation. All uses of the same switch tag can then share a single desugared tag. (*)

(As you have observed the other day, `cont.bind` can be expressed in terms of `switch`, which is the inverse encoding (which I haven't verified to work, but sounds plausible). But `cont.bind` is the simpler and more basic primitive.)

That all said, as is, the desugaring would still need to modify the tag type to add the continuation parameter. In that sense it becomes a different tag, and one could argue that it makes sense to treat it as such. Though perhaps a more elegant solution would be to require the tag type to actually have that very continuation parameter declared upfront instead of being empty, so that it really becomes fully equivalent.

(*) Btw, if we change the direction of cont.bind, then we should accordingly change the order of the parameters of the switch tag and its continuation, so that the continuation comes first. Otherwise this desugaring won't work, neither very likely use cases that need to use a similar trick. I'll open an issue for that.

Reply:
I guess the most important question was whether you would prefer some other semantics for the case when the same tag is used for multiple kinds of handlers simultaneously.

It's even easier to desugar `cont.bind` to a `resume-suspend` pair than to a `switch-switch` pair, and the best ordering of arguments for that is for `cont.bind` to take a suffix of the arguments of `resume` that includes the continuation, implying that the continuation comes last. You could even rewrite every `resume` that sends arguments into a sequence of `cont.bind`s that each binds a single argument followed by a nullary `resume` if you wanted to. More realistically, you could rewrite in the other direction: a `cont.bind` followed by a `resume` can be rewritten to just a `resume`. These rewritings of course require some transformations in continuation itself to account for the changing number and types of suspensions.